libs/api/ebpf_api.cpp

```diff
@@ -1464,28 +1469,38 @@ _delete_ebpf_object(_In_opt_ _Post_invalid_ ebpf_object_t* object) noexcept
     }
 }
 
-static void
-_remove_ebpf_object_from_globals(_In_ const ebpf_object_t* object) noexcept
+_Requires_lock_not_held_(_ebpf_state_mutex) static void _remove_ebpf_object_from_globals(
+    _In_ const ebpf_object_t* object) noexcept
 {
     EBPF_LOG_ENTRY();
     ebpf_assert(object);
+    std::unique_lock lock(_ebpf_state_mutex);
     auto it = std::find(_ebpf_objects.begin(), _ebpf_objects.end(), object);
     ebpf_assert(it != _ebpf_objects.end());
     _ebpf_objects.erase(it);
 }
 
-static void
-_clean_up_ebpf_objects() noexcept
+_Requires_lock_not_held_(_ebpf_state_mutex) static void _clean_up_ebpf_objects() noexcept
 {
     EBPF_LOG_ENTRY();
-    for (auto& object : _ebpf_objects) {
-        _delete_ebpf_object(object);
+    std::vector<ebpf_object_t*> objects;
+
+    // Intentional use of scope to limit lifetime of lock.
+    {
+        std::unique_lock lock(_ebpf_state_mutex);
+        objects = std::move(_ebpf_objects);
     }
 
-    _ebpf_objects.resize(0);
+    for (auto& object : objects) {
+        _delete_ebpf_object(object);
+    }
 
-    ebpf_assert(_ebpf_programs.size() == 0);
-    ebpf_assert(_ebpf_maps.size() == 0);
+    // Intentional use of scope to limit lifetime of lock.
+    {
+        std::unique_lock lock(_ebpf_state_mutex);
+        ebpf_assert(_ebpf_programs.size() == 0);
```

Comment:
Is `ebpf_assert` a `DEBUG` only macro? If so, perhaps you wrap this whole scope in `DEBUG` to prevent the unnecessary lock acquisition in release/fre builds for no reason.